packages/next-swc/crates/next-core/src/next_pages/page_entry.rs

```diff
@@ -34,116 +38,83 @@ pub async fn create_page_ssr_entry_module(
-                            page: {definition_page},
-                            pathname: {definition_pathname},
-                            // The following aren't used in production, but are
-                            // required for the RouteModule constructor.
-                            bundlePath: "",
-                            filename: "",
-                        }},
-                        components: {{
-                            App,
-                            Document,
-                        }},
-                        userland,
-                    }})
-                "#,
-                definition_page = StringifyJs(&definition_page),
-                definition_pathname = StringifyJs(&definition_pathname),
-            )?;
-
-            // When we're building the instrumentation page (only when the
-            // instrumentation file conflicts with a page also labeled
-            // /instrumentation) hoist the `register` method.
-            if definition_page.as_str() == "/instrumentation"
-                || definition_page.as_str() == "/src/instrumentation"
-            {
-                writeln!(
-                    result,
-                    r#"export const register = hoist(userland, "register")"#
-                )?;
-            }
-
-            "pages-entry.tsx".to_string()
+            // Load the Page entry file.
+            "/dist/esm/build/webpack/loaders/next-route-loader/entries/pages.js"
         }
         ReferenceType::Entry(EntryReferenceSubType::PagesApi) => {
-            // Sourced from https://github.com/vercel/next.js/blob/2848ce51d1552633119c89ab49ff7fe2e4e91c91/packages/next/src/build/webpack/loaders/next-route-loader/index.ts
-            writedoc!(
+            // Load the Pages API entry file.
+            "/dist/esm/build/webpack/loaders/next-route-loader/entries/pages-api.js"
+        }
+        _ => bail!("Invalid path type"),
+    };
+
+    // Load the file from the next.js codebase.
+    let file = load_next_js(project_root, template_file).await?.await?;
+
+    let mut file = file
+        .to_str()?
+        .replace(
+            "\"VAR_DEFINITION_PAGE\"",
+            &StringifyJs(&definition_page).to_string(),
+        )
+        .replace(
+            "\"VAR_DEFINITION_PATHNAME\"",
+            &StringifyJs(&definition_pathname).to_string(),
+        );
+
+    if reference_type == ReferenceType::Entry(EntryReferenceSubType::Page) {
+        file = file
+            .replace(
+                "\"VAR_MODULE_DOCUMENT\"",
+                &StringifyJs("@vercel/turbopack-next/pages/_document").to_string(),
+            )
+            .replace(
+                "\"VAR_MODULE_APP\"",
+                &StringifyJs("@vercel/turbopack-next/pages/_app").to_string(),
+            );
+    }
+
+    // Ensure that the last line is a newline.
+    if !file.ends_with('\n') {
+        file.push('\n');
+    }
+
+    let mut result = RopeBuilder::from(file.as_bytes().to_vec());
```

Comment:
Nit: there's a slight change we can make to avoid a second (internal) copy:


```rust
let file = Rope::from(file);
let mut result = RopeBuilder::default();
result += file
```